controllers/local_db_secret.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package controller
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	bs "janus-idp.io/backstage-operator/api/v1alpha1"
+)
+
+const (
+	_defaultPostGresSecretValue   = "sq4s3Eh4pw3N2"
+	postGresSecret                = "{POSTGRESQL_SECRET}"
+	_defaultPsqlMainContainerName = "postgresql"
+)
+
+func (r *BackstageReconciler) handlePsqlSecret(ctx context.Context, statefulSet *appsv1.StatefulSet, backstage *bs.Backstage) (*corev1.Secret, error) {
+	secretName := getSecretNameForGeneration(statefulSet, backstage)
+	if len(secretName) == 0 {
+		return nil, nil
+	}
+
+	var sec corev1.Secret
+	err := r.readConfigMapOrDefault(ctx, backstage.Spec.RawRuntimeConfig.BackstageConfigName, "db-secret.yaml", statefulSet.Namespace, &sec)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read config: %s", err)
+	}
+
+	//Generate the PostGresSQL secret if it does not exist
+	sec.SetName(secretName)
+	sec.SetNamespace(statefulSet.Namespace)
+	err = r.Get(ctx, types.NamespacedName{Name: secretName, Namespace: statefulSet.Namespace}, &sec)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return nil, fmt.Errorf("failed to get secret for PostgreSQL DB (%q), reason: %s", secretName, err)
+		}
+		// Create a secret with a random value
+		val := func(length int) string {
+			bytes := make([]byte, length)
+			if _, randErr := rand.Read(bytes); randErr != nil {
+				// Do not fail, but use a fallback value
+				return _defaultPostGresSecretValue
+			}
+			return string(bytes)
```

Comment:
In some cases, the generated password was not safe to use in the application, and made both the DB and Backstage pods unable to start.
After trying to decode the content in the Secret, I noticed that in some cases, we might end up with some unsafe characters, e.g.:

```
POSTGRES_PASSWORD: "\n\"��H��|W��\x03\x06uC\x05���K�i��"
```

```
$ kubectl get secrets backstage-psql-secret-bs1 -o jsonpath='{.data.POSTGRES_PASSWORD}' | base64 -d

"H|WuCKi%                                                                                                                                                                                      

```

or even:
```
POSTGRES_PASSWORD: "d옫�C\x18j��FT�w����ͦ��J�" 
```

```sh
# From the Backstage container
bash-5.1$ env | grep POSTGRES_PASS
POSTGRES_PASSWORD=d옫�Cj��FT�w����ͦ��J�
bash-5.1$ 
```

I think we'd better make it safer, maybe by further encoding it, or limiting the possible characters of the random generator to safe ones.

Reply:
Good suggestion! I added base64 encoding so that the generated password does not contain such special characters.